xmake/rules/c++/modules/modules_support/clang/builder.lua

```diff
@@ -0,0 +1,343 @@
+       if flags[2] then
+           table.join2(flags[2], {"-Wno-include-angled-in-module-purview", "-Wno-reserved-module-identifier"})
+       end
+    end
+
+    return table.unpack(flags)
+end
+
+-- get flags for building a headerunit
+function _make_headerunitflags(target, headerunit, bmifile)
+
+    local module_headerflag = compiler_support.get_moduleheaderflag(target)
+    assert(module_headerflag, "compiler(clang): does not support c++ header units!")
+
+    local local_directory = (headerunit.type == ":quote") and {"-I" .. path.directory(headerunit.path)} or {}
+
+    local headertype = (headerunit.type == ":angle") and "system" or "user"
+
+    local flags = table.join(local_directory, {"-xc++-header", "-Wno-everything", module_headerflag .. headertype, "-o", bmifile, "-c", headerunit.path})
+
+    return flags
+end
+
+-- do compile
+function _compile(target, flags, cppfile)
+
+    local dryrun = option.get("dry-run")
+    local compinst = target:compiler("cxx")
+    local compflags = compinst:compflags({sourcefile = cppfile, target = target})
+    local msvc = target:toolchain("msvc")
+    local flags = table.join(compflags or {}, flags)
+
+    -- trace
+    vprint(compinst:program(), table.unpack(flags))
+
+    if not dryrun then
+        -- do compile
+        if msvc then
+            local _, err = os.iorunv(compinst:program(), flags, {envs = msvc:runenvs()})
+            assert(err, err)
+        else
+            local _, err = os.iorunv(compinst:program(), flags)
+            assert(err, err)
+        end
+    end
+end
+
+-- do compile for batchcmds
+-- @note we need to use batchcmds:compilev to translate paths in compflags for generator, e.g. -Ixx
+function _batchcmds_compile(batchcmds, target, flags, sourcefile)
+
+    local compinst = target:compiler("cxx")
+    local compflags = compinst:compflags({sourcefile = sourcefile, target = target})
+    batchcmds:compilev(table.join(compflags or {}, flags), {compiler = compinst, sourcekind = "cxx"})
+end
+
+-- get module requires flags
+-- e.g
+-- -fmodule-file=build/.gens/Foo/rules/modules/cache/foo.pcm
+-- -fmodule-file=build/.gens/Foo/rules/modules/cache/iostream.pcm
+-- -fmodule-file=build/.gens/Foo/rules/modules/cache/bar.hpp.pcm
+-- on LLVM >= 16
+-- -fmodule-file=foo=build/.gens/Foo/rules/modules/cache/foo.pcm
+-- -fmodule-file=build/.gens/Foo/rules/modules/cache/iostream.pcm
+-- -fmodule-file=build/.gens/Foo/rules/modules/cache/bar.hpp.pcm
+--
+function _get_requiresflags(target, module, opt)
+
+    local modulefileflag = compiler_support.get_modulefileflag(target)
+
+    local name = module.name
+    local cachekey = target:name() .. name
+
+    local requiresflags = compiler_support.memcache():get2(cachekey, "requiresflags")
+                       or compiler_support.localcache():get2(cachekey, "requiresflags")
+
+    if not requiresflags or (opt and opt.regenerate) then
+        requiresflags = {}
+        for required, _ in pairs(module.requires) do
```

Comment:
we should use `table.orderpairs(requires)` to ensure the flags detection for incremental compilation does not change

https://github.com/xmake-io/xmake/blob/1d4a3e7645143f98ea1089b63153834f7c913d4b/xmake/rules/c%2B%2B/modules/modules_support/clang.lua#L952